llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -853,6 +853,16 @@ NVPTXTargetLowering::NVPTXTargetLowering(const NVPTXTargetMachine &TM,
       AddPromotedToType(Op, MVT::bf16, MVT::f32);
   }
 
+  // Lower bf16 add/mul/sub as fma when it avoids promotion
+  for (const auto &Op : {ISD::FADD, ISD::FMUL, ISD::FSUB}) {
+    for (const auto &VT : {MVT::bf16, MVT::v2bf16}) {
+      if (getOperationAction(Op, VT) != Legal &&
+          getOperationAction(ISD::FMA, VT) == Legal) {
```

Comment:
This is cumbersome, we usually don't write legalizer rules in terms of other legalizer rules.

I think you'd be best off just putting this logic into the default Expand action for add/fmul/fsub. If the FMA is legal, you emit the appropriate sequence before falling back to the default libcall expansion. Then you shouldn't need to touch the target rules here

Reply:
> This is cumbersome, we usually don't write legalizer rules in terms of other legalizer rules.

I could easily write it in terms of SM and PTX version ranges instead, I just felt this more directly expressed the rationale behind the version range.

> I think you'd be best off just putting this logic into the default Expand action for add/fmul/fsub. If the FMA is legal, you emit the appropriate sequence before falling back to the default libcall expansion. Then you shouldn't need to touch the target rules here

I'm not sure this makes sense. The FTZ logic is target specific and we also want to fallback to promotion, not a libcall here.

Reply:
> I'm not sure this makes sense. The FTZ logic is target specific and we also want to fallback to promotion, not a libcall here.

The custom expansion can defer to the default expansion depending on the function state. Yes, the default Expand action can conditionally use the to FMA path instead of the default libcall expansion. The core transform code can be in the generic legalizer

Reply:
Alright, done.